SLING-11609: Provide doGetWithRetry methods in Sling Testing Clients

nscendoni opened the pull request:
In several situations during IT Tests development we need to retry doGet method until some assertion are verified.

I propose in the attached pull request an implementation and test for new methods doGetWithRetry

joerghoh commented on src/main/java/org/apache/sling/testing/clients/AbstractSlingClient.java:
```diff
@@ -539,6 +541,36 @@ public SlingHttpResponse doGet(String requestPath, int... expectedStatus)
+            p.poll(timeout, delay);
+            return p.getResponse();
+        } catch (TimeoutException | InterruptedException e) {
+            throw new TestingValidationException(e.getMessage(), e);
```
is the TestingValidationException really the right exception to throw here? It would suggest that there was a result which did not match the expectation (e.g. a 200 instead of a 404).
I would suggest to rather throw a TestingIOException,

nscendoni replied:
Sonarclub check also reported that I should not catch and raise InterruptedException. I'm not catching them anymore. Is that ok?

joerghoh commented on src/main/java/org/apache/sling/testing/clients/AbstractSlingClient.java:
```diff
@@ -733,4 +765,46 @@ public <T> T execute(HttpHost target, HttpRequest request, ResponseHandler<? ext
+                        this.response = response;
+                        return true;
+                    } catch (TimeoutException | InterruptedException | AssertionError e) {
+                        throw new TestingValidationException(e.getMessage(), e);
```
Please throw the TestingValidationException only when Assertion failed.

nscendoni replied:
I'm not catching TimeoutException and  InterruptedException  anymore.

anchela commented on src/main/java/org/apache/sling/testing/clients/AbstractSlingClient.java:
```diff
@@ -733,4 +761,46 @@ public <T> T execute(HttpHost target, HttpRequest request, ResponseHandler<? ext
     }
+
+    private class GetPolling extends Polling {
+            private int lastStatusCode = -1;
```
i believe 'lastStatusCode' could be a local field in the method

anchela commented on src/main/java/org/apache/sling/testing/clients/AbstractSlingClient.java:
```diff
@@ -733,4 +761,46 @@ public <T> T execute(HttpHost target, HttpRequest request, ResponseHandler<? ext
+            private List<Header> headers;
+            private ResponseAssertion assertion;
+
+        private SlingHttpResponse response;
```
inconsistent indention.... (being picky)

anchela commented on src/main/java/org/apache/sling/testing/clients/AbstractSlingClient.java:
```diff
@@ -733,4 +761,46 @@ public <T> T execute(HttpHost target, HttpRequest request, ResponseHandler<? ext
+            public Boolean call() throws ClientException, InterruptedException, TimeoutException {
+            SlingHttpResponse response = doGet(requestPath, parameters, headers);
+            lastStatusCode = response.getStatusLine().getStatusCode();
+            for (int count=0; count<expectedStatus.length; count++) {
```
that could imho be using the enhanced for-loop: `for (int status : expectedStatus)`

anchela reviewed (commented):
hi @nscendoni , just added a few minor nitpicking comments..... for the rest of the super minor stuff see the patch....

nscendoni commented:
@enapps-enorman Thanks for the suggestion. awaitility is a very good good library I didn't know.
Unfortunately I had to implement anyway a private class in order to return `SlingHttpResponse`

@joerghoh: Exceptions now are managed by awaitility
@anchela: Thanks for the review. I incorporated the suggestions on the new code.

anchela reviewed (approved):
@nscendoni , the changes look good to me. but take it with a grain of salt as i am not particularly familiar with the testing clients code base.

enapps-enorman commented:
@nscendoni If I am not mistaken, I believe using Awaitility apis along with the existing SlingClient#doGet call could accomplish the same thing without introducing any new code into the Sling Testing Clients.

For example, something like this:
`    @Test
    public void testDoGetWithRetry() throws ClientException, InterruptedException, TimeoutException {
        SlingClient c = new SlingClient(httpServer.getURI(), "user", "pass");

        Awaitility.await("doGet until expectedStatus and Assertions are satisfied")
            .atMost(Duration.ofSeconds(1000))
            .pollInterval(Duration.ofMillis(100))
            .ignoreExceptionsInstanceOf(ClientException.class)
            .until(() -> {
                SlingHttpResponse response = c.doGet("/", 200);
                return SUCCESS_RESPONSE.equals(response.getContent());
            });
    }
`
Does that make sense?  Do you have other requirements that need something else?

nscendoni commented:
@enapps-enorman Yes, I can use Awaitility without modifying sling-testing-clients and the code is compact and clean. 
You can either close this pull request or merge it if you think this could be useful to other people.

enapps-enorman commented:
> I wouldn't mind if we keep this functionality, even if it is just a thin wrapper around `Awaitility`.

@joerghoh I won't stand in the way of that since this just just for testing, but it seems redundant and the solution is incomplete since it is only handling the 'GET' request scenario.  The same problem could arise for a head, post, put, delete, patch or stream action as well.  The solution I proposed would handle all other scenarios the same way.

Perhaps a simple example in the [readme](https://github.com/apache/sling-org-apache-sling-testing-clients/blob/master/README.md) FAQ would suffice?

joerghoh commented:
@enapps-enorman It's fine for me if we move the example you have given above into the readme. It's just that I don't want to loose that snippet here in the github PR section.

@nscendoni would you mind to change the PR and add it to the README instead?

nscendoni commented:
Hi @joerghoh
I reverted my changes and added a small section in the Readme about how to retry HTTP Requests including the coded snippet provided by @enapps-enorman
